processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.internal.model.source.selector;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.mapstruct.ap.internal.model.common.Parameter;
+import org.mapstruct.ap.internal.model.common.ParameterBinding;
+import org.mapstruct.ap.internal.model.source.Method;
+
+/**
+ * Selector for deduplicating overloaded lifecycle callback methods
+ * whose parameter signatures differ only by type hierarchy.
+ * <p>
+ * In the context of lifecycle callback method selection
+ * (such as @BeforeMapping or @AfterMapping), it is possible to have multiple overloaded methods
+ * whose parameter lists are structurally identical except for the specific types,
+ * where those types are related by inheritance (e.g., one parameter is a superclass or subclass of another).
+ * <p>
+ * This selector groups such methods by their effective parameter signature
+ * (ignoring differences only in type hierarchy), and, within each group,
+ * retains only the method whose parameter types have the closest inheritance distance
+ * to the actual invocation types.
+ * This ensures that, for each group of nearly identical overloads,
+ * only the most specific and appropriate method is selected.
+ * <p>
+ * <b>Example (see Issue3849Test):</b>
+ *
+ * @AfterMapping default void afterMapping(Parent source, @MappingTarget ParentDto target) { ... }
+ * @AfterMapping default void afterMapping(Parent source, @MappingTarget ChildDto target) { ... }
+ * </pre>
+ * When mapping a Child to a ChildDto,
+ * only the method with ChildDto is selected, even though both methods match by signature
+ * except for the target type's inheritance relationship.
+ */
+public class LifecycleOverloadDeduplicateSelector implements MethodSelector {
+    @Override
+    public <T extends Method> List<SelectedMethod<T>> getMatchingMethods(List<SelectedMethod<T>> methods,
+                                                                         SelectionContext context) {
+        if ( !context.getSelectionCriteria().isLifecycleCallbackRequired() ) {
+            return methods;
+        }
+        List<List<SelectedMethod<T>>> methodSignatureGroups = new ArrayList<>(
+            methods.stream()
+                .collect( Collectors.groupingBy(
+                    LifecycleOverloadDeduplicateSelector::buildSignatureKey,
+                    LinkedHashMap::new,
+                    Collectors.toList()
+                ) )
+                .values()
```

Comment:
Let's try to avoid creating a lot of objects. There is no need to create a new list for the values. You can just get the values (`Collection<List<SelectedMethod<T>>>`) and iterate over it